google-ads-examples/src/main/java/com/google/ads/googleads/examples/misc/NavigateSearchResultPagesCachingTokens.java

```diff
@@ -0,0 +1,247 @@
+      // Instances of this exception have a message and a GoogleAdsFailure that contains a
+      // collection of GoogleAdsErrors that indicate the underlying causes of the
+      // GoogleAdsException.
+      System.err.printf(
+          "Request ID %s failed due to GoogleAdsException. Underlying errors:%n",
+          gae.getRequestId());
+      int i = 0;
+      for (GoogleAdsError googleAdsError : gae.getGoogleAdsFailure().getErrorsList()) {
+        System.err.printf("  Error %d: %s%n", i++, googleAdsError);
+      }
+      System.exit(1);
+    }
+  }
+
+  /**
+   * Runs this example.
+   *
+   * @param googleAdsClient the Google Ads API client.
+   * @param customerId the client customer ID.
+   * @throws GoogleAdsException if an API request failed with one or more service errors.
+   */
+  private static void runExample(GoogleAdsClient googleAdsClient, long customerId) {
+    // The cache of page tokens. It is stored in-memory and in ascendant order of page number.
+    // The first page's token is always an empty string.
+    List<String> pageTokens = new ArrayList<>(Collections.singletonList(""));
+
+    System.out.println("--- 0. Fetch page 1 to get metadata:");
+
+    // Creates a query that retrieves the campaigns.
+    String query =
+        String.format(
+            "SELECT campaign.id, campaign.name FROM campaign ORDER BY campaign.name LIMIT %d",
+            RESULTS_LIMIT);
+
+    // Creates a paginated search request.
+    SearchGoogleAdsRequest request =
+        SearchGoogleAdsRequest.newBuilder()
+            .setCustomerId(Long.toString(customerId))
+            .setPageSize(PAGE_SIZE)
+            .setQuery(query)
+            .setReturnTotalResultsCount(true)
+            .build();
+
+    int totalNumberOfPages;
+    try (GoogleAdsServiceClient googleAdsServiceClient =
+        googleAdsClient.getLatestVersion().createGoogleAdsServiceClient()) {
+      // Issues a paginated search request.
+      SearchPagedResponse response = googleAdsServiceClient.search(request);
+      cacheNextPageToken(pageTokens, response.getPage(), 1);
+
+      // Determines the total number of results and prints it.
+      // The total results count does not take into consideration the LIMIT clause of the query,
+      // so we need to find the minimal value between the limit and the total results count.
+      long totalNumberOfResults =
+          Math.min(RESULTS_LIMIT, response.getPage().getResponse().getTotalResultsCount());
+      System.out.printf("Total number of campaigns found: %d.%n", totalNumberOfResults);
+
+      // Determines the total number of pages and prints it.
+      totalNumberOfPages = (int) Math.ceil((double) totalNumberOfResults / (double) PAGE_SIZE);
+      System.out.printf("Total number of pages: %d.%n", totalNumberOfPages);
+      if (totalNumberOfPages == 0) {
+        System.out.println("Could not find any campaigns.");
+        System.exit(1);
```

Comment:
Prefer `return` over `System.exit`. We should only do the latter in `main` so if someone copies and pastes this code they won't inadvertently shut down their JVM on an empty response.

If detecting an empty response is important (e.g., for FT), then you could just throw an exception here such as `IllegalStateException`.